### DISCUSSION TITLE
Avoiding defensive copy in custom binary types

### DISCUSSION THREAD

**Discussion by DmitryMak** [nhibernate/nhibernate-core#3540] [Q&A]
It looks like NHibernate requires bytes set in IUserType.NullSafeSet to be a Byte array. Sometimes the underlying Byte array is not available and we only have read only collection or enumerable. So the only option is to perform a defensive copy which can be expensive for large arrays:

```
public class MyGizmoType : IUserType {
    
    public SqlType[] SqlTypes {
        get { return [new(DbType.Binary)]; }
    }
    
    public void NullSafeSet(DbCommand cmd, Object value, Int32 index, ISessionImplementor session) {
        var parameter = cmd.Parameters[index];
        
        if (value == null) {
            parameter.Value = DBNull.Value;
        } else {

            var gizmo = (Gizmo)value;
            
            // gizmo.ReadOnlyBytes is IReadOnlyList<Byte> or IEnumerable<Byte>

            parameter.Value = gizmo.ReadOnlyBytes.ToArray(); // <- How can I avoid this large defensive copy?
        }
    }
} 
```

**Comment by trivalik** [comment]
Maybe with an abstraction object which provide you the readonly behavior, the entity for database can be read and writeable. Then no custom type would be necessary.

Alternative would be when `ReadOnlyBytes` are castable to array :-)
